iceoryx_hoofs/include/iceoryx_hoofs/internal/log/platform_building_blocks/logger.inl

```diff
@@ -0,0 +1,137 @@
+
+#include <atomic>
+#include <cstdint>
+#include <cstring>
+#include <mutex>
+
+namespace iox
+{
+namespace pbb
+{
+template <uint32_t N>
+// NOLINTJUSTIFICATION see at declaration in header
+// NOLINTNEXTLINE(hicpp-avoid-c-arrays, cppcoreguidelines-avoid-c-arrays)
+inline constexpr bool equalStrings(const char* lhs, const char (&rhs)[N]) noexcept
+{
+    return strncmp(lhs, rhs, N) == 0;
+}
+
+
+template <typename BaseLogger>
+inline constexpr LogLevel Logger<BaseLogger>::minimalLogLevel() noexcept
+{
+    return Logger::MINIMAL_LOG_LEVEL;
+}
+
+template <typename BaseLogger>
+inline constexpr bool Logger<BaseLogger>::ignoreLogLevel() noexcept
+{
+    return Logger::IGNORE_ACTIVE_LOG_LEVEL;
+}
+
+
+template <typename BaseLogger>
+inline Logger<BaseLogger>& Logger<BaseLogger>::get() noexcept
+{
+    // NOLINTJUSTIFICATION needed for the functionality
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    thread_local static Logger* logger = Logger::activeLogger();
+    if (!logger->m_isActive.load(std::memory_order_relaxed))
+    {
+        // no need to loop until m_isActive is true since this is an inherent race
+        //   - the logger needs to be active for the whole lifetime of the application anyway
+        //   - if the logger was changed again, the next call will update the logger
+        //   - furthermore, it is not recommended to change the logger more than once
+        logger = Logger::activeLogger();
+    }
+    return *logger;
+}
+
+template <typename BaseLogger>
+inline void Logger<BaseLogger>::init(const LogLevel logLevel) noexcept
+{
+    Logger::get().initLoggerInternal(logLevel);
+}
+
+template <typename BaseLogger>
+inline void Logger<BaseLogger>::setActiveLogger(Logger<BaseLogger>* newLogger) noexcept
+{
+    Logger::activeLogger(newLogger);
+}
+
+template <typename BaseLogger>
+inline Logger<BaseLogger>* Logger<BaseLogger>::activeLogger(Logger<BaseLogger>* newLogger) noexcept
+{
+    static std::mutex mtx;
+    std::lock_guard<std::mutex> lock(mtx);
+    static Logger defaultLogger;
+
+    // NOLINTJUSTIFICATION needed for the functionality
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    static Logger* logger{&defaultLogger};
+
+    if (newLogger)
+    {
+        if (logger->m_isFinalized.load(std::memory_order_relaxed))
+        {
+            logger->createLogMessageHeader(__FILE__, __LINE__, __FUNCTION__, LogLevel::ERROR);
+            logger->logString("Trying to replace logger after already initialized!");
+            logger->flush();
+            newLogger->createLogMessageHeader(__FILE__, __LINE__, __FUNCTION__, LogLevel::ERROR);
+            logger->logString("Trying to replace logger after already initialized!");
+            newLogger->flush();
+            /// @todo iox-#1345 call error handler after the error handler refactoring was merged
+        }
+        else
+        {
+            logger->m_isActive.store(false);
```

Comment:
OK, assume T1 sets it to false but has not set the new logger yet. T2 reads that it is false and sets the new logger to the old one (which is still inactive). It may do this multiple times until it finally sets the new logger. This is fine.

If we switch the statement order it will just not have to retry, but this is of not much value so it is fine as is. Besides we do not want to store the old logger. There may also be other issues if we switch this.

Reply:
The logger is expected to only be set once and then the order is not important since everything is guarded by a mutex.